### PR TITLE
`app-tests`: validates HTML source saved from page-layout tests

### LINKED ISSUE #1897: Lint HTML
To catch issues like the [missing](https://github.com/freedomofpress/securedrop/pull/1829/commits/05e5ddb05d96f2a107410d1c6e5fa63dcf286d72#r122849978) `</a>` tag found in #1829, we should find a good linter for HTML (e.g. [html-linter](https://pypi.python.org/pypi/html-linter/)), and add it to CI.

### PR DESCRIPTION
**Pull request by cfm**
## Status

Ready for review

## Description of Changes

Closes #1897 by:

1. installing [`html5validator`](https://github.com/svenkreiss/html5validator) in the `test-requirements` used in CI pipelines;
2. providing and using in all page-layout tests a convenience function `FunctionalTest._save_html()` in the spirit of `FunctionalTest._screenshot()`;
3. adding a Make target `validate-test-html` to validate the HTML source saved from such tests; and
4. executing this target (as informational-only for now) after tests have run in `app-tests`.

### Discussion points

1. This is arguably a strict form of linting, but (per discussion in #1897) requires the outputs of higher-level tests, in this case page-layout tests.  Given this infelicity, is there some other way we want to integrate this check into CI?  For example, we could run it only as part of the `nightly` workflow.
2. If we do want this to be part of routine CI:  Do we want to include this step in `app-tests`, further prolonging its runtime?  One alternative would be to factor out page-layout tests to a separate job altogether, in parallel to `app-tests`, and include this validation step there.
3. `html5validator` requires a JDK, installed here as `default-jdk`.  Is this an acceptable addition to the development/test container?

## Testing

- [ ] After running page-layout tests locally, `make validate-test-html` reports meaningful results (i.e., lots of errors :-).
- [ ] `app-tests` passes ([e.g.](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2861/workflows/cdfcc0af-3a13-4bc7-813c-48a2901d3ba5/jobs/56177)) despite lots of "informational" errors reported in the "Validate HTML" step.

## Deployment

Development-only; no deployment considerations.

### If you added or updated a code dependency:

- ~I have performed a diff review and pasted the contents to [the packaging wiki](https://github.com/freedomofpress/securedrop-debian-packaging/wiki)~
- ~I would like someone else to do the diff review~

`test-requirements` only.

### CODE REVIEW

File: securedrop/dockerfiles/focal/python3/Dockerfile [hunk 1]
```diff
@@ -13,6 +13,8 @@ RUN apt-get update && apt-get install -y paxctl && \
                        python3-pip python3-all python3-venv virtualenv libpython3.8-dev libssl-dev \
                        gnupg2 ruby redis-server git xvfb curl wget \
                        gettext paxctl x11vnc enchant libffi-dev sqlite3 gettext sudo \
+                       # for html5validator
+                       default-jdk \
```
Comment (kushaldas) [inline securedrop/dockerfiles/focal/python3/Dockerfile:17]: Finally we are an enterprise application :)

**Review by kushaldas — CHANGES_REQUESTED** [review]
The code and rest of the change looks good to me.  Tested locally, can get many errors like below:

```
"file:/root/code/securedrop/securedrop/tests/pageslayout/html/ar/journalist-admin_ossec_alert_button.html":1.1-1.26: error: Start tag seen without seeing a doctype first. Expected "<!DOCTYPE html>".
"file:/root/code/securedrop/securedrop/tests/pageslayout/html/ar/journalist-admin_ossec_alert_button.html":36.5-36.37: error: Element "div" not allowed as child of element "label" in this context. (Suppressing further errors from this subtree.)
"file:/root/code/securedrop/securedrop/tests/pageslayout/html/ar/journalist-admin_ossec_alert_button.html":39.5-39.37: error: Element "div" not allowed as child of element "label" in this context. (Suppressing further errors from this subtree.)
"file:/root/code/securedrop/securedrop/tests/pageslayout/html/ar/journalist-admin_ossec_alert_button.html":42.5-42.40: error: Element "div" not allowed as child of element "label" in this context. (Suppressing further errors from this subtree.)
"file:/root/code/securedrop/securedrop/tests/pageslayout/html/ar/journalist-login.html":1.1-1.26: error: Start tag seen without seeing a doctype first. Expected "<!DOCTYPE html>".
"file:/root/code/securedrop/securedrop/tests/pageslayout/html/ar/journalist-login.html":30.5-30.37: error: Element "div" not allowed as child of element "label" in this context. (Suppressing further errors from this subtree.)
"file:/root/code/securedrop/securedrop/tests/pageslayout/html/ar/journalist-login.html":33.5-33.37: error: Element "div" not allowed as child of element "label" in this context. (Suppressing further errors from this subtree.)
"file:/root/code/securedrop/securedrop/tests/pageslayout/html/ar/journalist-login.html":36.5-36.40: error: Element "div" not allowed as child of element "label" in this context. (Suppressing further errors from this subtree.)
make: *** [Makefile:231: validate-test-html] Error 255
```

But, in CI I can not see the actual errors. I guess we want that too (as mentioned in the test plan).

image


@SaptakS please let me know what do you think.

**Comment by cfm** [comment]
@kushaldas in https://github.com/freedomofpress/securedrop/pull/6072#pullrequestreview-732669178:

> But, in CI I can not see the actual errors. I guess we want that too (as mentioned in the test plan).

This threw me too!  As CircleCI is currently splitting the tests across the three parallel runs, run 0 (at least on this execution) doesn't include any page-layout tests, so it doesn't save any HTML to validate.  But runs [1](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2861/workflows/cdfcc0af-3a13-4bc7-813c-48a2901d3ba5/jobs/56177/parallel-runs/1/steps/1-110) and [2](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2861/workflows/cdfcc0af-3a13-4bc7-813c-48a2901d3ba5/jobs/56177/parallel-runs/2/steps/2-110) do.

I tried to anticipate this by having the `validate-test-html` target start by echoing "Validating HTML source saved **from any page-layout tests run**" (emphasis added).  But I will think about whether there's a better way to document this behavior and/or highlight it in the CircleCI view.

**Comment by cfm** [comment]
https://github.com/freedomofpress/securedrop/pull/6072#issuecomment-901209493:

> I tried to anticipate this by having the `validate-test-html` target start by echoing "Validating HTML source saved **from any page-layout tests run**" (emphasis added). But I will think about whether there's a better way to document this behavior and/or highlight it in the CircleCI view.

In 2ed9fce100aa10fc1382eb491ee71f70045fbe44 I've added an explicit count of how many `*.html` files have been saved from page-layout tests and will be validated.

**Review by kushaldas — APPROVED** [review]
## Testing steps


- [x] After running page-layout tests locally, `make validate-test-html` reports meaningful results (i.e., lots of errors :-).
- [x] `app-tests` passes ([e.g.](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2861/workflows/cdfcc0af-3a13-4bc7-813c-48a2901d3ba5/jobs/56177)) despite lots of "informational" errors reported in the "Validate HTML" step.

Tested locally, works as expected. `html` directory is also marked in `.gitignore`.

**Pull request merged**